### PR TITLE
Add BERT model and training script

### PR DESCRIPTION
**Pull request by amkatrutsa**

### CODE REVIEW

**Comment by daskol** [comment]
BERT and RoBERTa are my part. Are you going finish BERT or I am proceeding from now on.

**Comment by Muxas** [comment]
> BERT and RoBERTa are my part. Are you going finish BERT or I am proceeding from now on.

You are not even visiting our team chat. Your participation in the project is under a big question. I have to write this in a public space because there is no private way to contact you.

File: wrappers/python/nntile/model/bert_config.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    num_attention_heads: int
+    n_head_tile: int
+    activation_function: str = "gelutanh"
+    dtype: str = "fp32"
+    flashattention: bool = False
+    layer_norm_epsilon: float = 1e-5
+    max_position_embeddings: int = 512
+    num_hidden_layers: int = 1
+    redux: bool = True
+    hidden_dropout_prob = 0
```
Comment (Muxas) [inline wrappers/python/nntile/model/bert_config.py:34]: `prob: float = 0.0` to denote type

File: wrappers/python/nntile/model/bert_config.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    attention_probs_dropout_prob = 0
```
Comment (Muxas) [inline wrappers/python/nntile/model/bert_config.py:35]: type

File: wrappers/python/nntile/model/bert_config.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    classifier_dropout = 0
```
Comment (Muxas) [inline wrappers/python/nntile/model/bert_config.py:36]: type

File: wrappers/python/examples/masked_lm_data_preparation.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+# and move our model over to the selected device
+model.to(device)
+print(device)
+
+# activate training mode
+model.train()
+# initialize optimizer
+optim = AdamW(model.parameters(), lr=1e-4)
+
+epochs = 3
```
Comment (Muxas) [inline wrappers/python/examples/masked_lm_data_preparation.py:83]: script name says data preparation, but the script is about data preparation AND training

File: wrappers/python/tests/model/test_bert_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        torch_model, nntile_model, x, y_grad = generate_inputs(params, dtype)
+        y = torch_model(x)
+        nntile_model.forward_async()
+        res = (y * y_grad).sum()
+        res.backward()
+
+        nntile_model.backward_async()
+        torch_model_other = nntile_model.to_torch_with_grads()
+        nntile_model.unregister()
+
```
Comment (Muxas) [inline wrappers/python/tests/model/test_bert_embeddings.py:202]: check gradient over input also

**Review by Muxas — COMMENTED** [review]
So far it seems good. Proceed with other BERT blocks

File: wrappers/python/tests/model/test_bert_embeddings.py (hunk 5 above)
Comment (amkatrutsa) [reply wrappers/python/tests/model/test_bert_embeddings.py:202]: Gradients for the int64 variables look strange

File: wrappers/python/examples/masked_lm_data_preparation.py (hunk 4 above)
Comment (amkatrutsa) [reply wrappers/python/examples/masked_lm_data_preparation.py:83]: This script is just fixing the current version of the general pipeline, it will be revised in future commits

File: wrappers/python/tests/model/test_bert_embeddings.py (hunk 5 above)
Comment (Muxas) [reply wrappers/python/tests/model/test_bert_embeddings.py:202]: indeed

File: wrappers/python/nntile/layer/bert_selfattention.py [hunk 6]
```diff
@@ -0,0 +1,1053 @@
...
+        self.head_size = head_size
+        self.mask = mask
+        if mask:
+            self.val = -np.float32(np.inf)
+        if redux:
+            self.redux = 1
+        else:
+            self.redux = 0
+
+    # Simple generator for the linear layer
```
Comment (Muxas) [inline wrappers/python/nntile/layer/bert_selfattention.py:182]: remove "for the linear layer"

File: wrappers/python/tests/model/test_bert_output.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    pytest.param(multiple_tiles, id='multiple_tiles'),
+])
+@pytest.mark.parametrize('dtype', [
+    'fp32',
+    pytest.param('bf16', marks=nocuda),
+    pytest.param('fp32_fast_tf32', marks=nocuda),
+    pytest.param('fp32_fast_fp16', marks=nocuda),
+    pytest.param('fp32_fast_bf16', marks=nocuda),
+])
+class TestBertEmbeddings:
```
Comment (Muxas) [inline wrappers/python/tests/model/test_bert_output.py:174]: TestBertOutput

File: src/kernel/subtract_indexed_outputs/cpu.cc [hunk 8]
```diff
@@ -46,24 +48,29 @@ void cpu(Index n_labels, Index n_outputs, Scalar val_, const int64_t* labels_,
     Y dst_val{0.0};
     for(Index i = 0; i < n_outputs; ++i)
     {
-        dst_val = static_cast<Y>(dst_[labels[i] + i*n_labels]);
-        dst_[labels[i] + i*n_labels] = static_cast<T>(dst_val - val);
+        if (labels[i] == ignore_index)
+        {
+            std::memset(dst_ + i*n_labels, 0, n_labels * sizeof(T));
```
Comment (Muxas) [inline src/kernel/subtract_indexed_outputs/cpu.cc:53]: If a label shall be ignored, its gradient shall be untouched. In the beginning all gradients are zeroed and some operations might update them. If `subtract_indexed_outputs` is called after another function that actually updates gradient, it will destroy already accumulated gradient. And that is a bug. There shall be no update in case a label is to be ignored.

File: src/kernel/subtract_indexed_outputs/cuda.cu [hunk 9]
```diff
@@ -20,21 +20,26 @@ namespace nntile::kernel::subtract_indexed_outputs
...
+                 Scalar val_, const Index* labels, T *dst)
 {
     int i = threadIdx.x + blockIdx.x*blockDim.x;
     using Y = typename T::repr_t;
     Y dst_val{0.0};
     Y val{val_};
     if(i < n_outputs)
     {
+        if (labels[i] == ignore_index) {
+            ::memset(dst + i*n_labels, 0, n_labels * sizeof(T));
```
Comment (Muxas) [inline src/kernel/subtract_indexed_outputs/cuda.cu:33]: there shall be no update if a label is to be ignored

File: src/kernel/total_sum_accum/cuda.cu [hunk 10]
```diff
@@ -20,32 +20,35 @@ namespace nntile::kernel::total_sum_accum
...
         if(threadIdx.x == 0)
         {
             block_val = 0;
         }
-        float val1 = static_cast<Y>(logsumexp[i]);
-        float val2 = static_cast<Y>(src[labels[i] + i*n_labels]);
+        Y val1 = static_cast<Y>(logsumexp[i]);
+        Y val2 = val1;
+        if (labels[i] != ignore_index)
+            val2 = static_cast<Y>(src[labels[i] + i*n_labels]);
```
Comment (Muxas) [inline src/kernel/total_sum_accum/cuda.cu:38]: move `atomicAdd` to the inside of this if statement. if a label is to be ignored, it atomicAdd will just waste GPU cycles to resolve `atomicAdd` data synchronizations.

File: src/starpu/subtract_indexed_outputs.cc [hunk 11]
```diff
@@ -29,13 +29,15 @@ void cpu(void *buffers[], void *cl_args)
...
+    Scalar val = args->value;
     // Get interfaces
     auto interfaces = reinterpret_cast<VariableInterface **>(buffers);
     const int64_t *labels = interfaces[0]->get_ptr<int64_t>();
     T *dst = interfaces[1]->get_ptr<T>();
     // Launch kernel
-    kernel::subtract_indexed_outputs::cpu<T>(n_labels, n_outputs, args->value,
-            labels, dst);
+    kernel::subtract_indexed_outputs::cpu<T>(n_labels, n_outputs, ignore_index,
+                                             val, labels, dst);
```
Comment (Muxas) [inline src/starpu/subtract_indexed_outputs.cc:40]: 8 spaces to make intendation the same across this file

File: src/starpu/total_sum_accum.cc [hunk 12]
```diff
@@ -65,7 +67,8 @@ void cuda(void *buffers[], void *cl_args)
     cudaStream_t stream = starpu_cuda_get_local_stream();
     // Launch kernel
     kernel::total_sum_accum::cuda<T>(stream, alpha, n_labels, n_outputs,
-            logsumexp, src, labels, val);
+                                     ignore_index, logsumexp, src, labels,
```
Comment (Muxas) [inline src/starpu/total_sum_accum.cc:70]: 8 spaces for the intendation

File: wrappers/python/nntile/functions.py [hunk 13]
```diff
@@ -1467,56 +1467,61 @@ def total_sum_accum_async(
...
 
 def subtract_indexed_outputs_async(
-    val: float, class_labels: Tensor_int64, dst: Tensor
+    val: float, class_labels: Tensor_int64, dst: Tensor,
+    ignore_index: int
 ):
     if type(dst) is core_tensor.Tensor_fp32:
-        core_tensor.subtract_indexed_outputs_async_fp32(val, class_labels, dst)
+        core_tensor.subtract_indexed_outputs_async_fp32(val, class_labels, dst,
+                                                        ignore_index)
```
Comment (Muxas) [inline wrappers/python/nntile/functions.py:1506]: use the same style to provide arguments

File: wrappers/python/nntile/functions.py [hunk 14]
```diff
@@ -1467,56 +1467,61 @@ def total_sum_accum_async(
...
+            val, class_labels, dst, ignore_index
         )
     elif type(dst) is core_tensor.Tensor_fp32_fast_bf16:
         core_tensor.subtract_indexed_outputs_async_fp32_fast_bf16(
-            val, class_labels, dst
+            val, class_labels, dst, ignore_index
         )
     elif type(dst) is core_tensor.Tensor_fp64:
-        core_tensor.subtract_indexed_outputs_async_fp64(val, class_labels, dst)
+        core_tensor.subtract_indexed_outputs_async_fp64(val, class_labels,
```
Comment (Muxas) [inline wrappers/python/nntile/functions.py:1520]: use the same style to provide args

File: wrappers/python/nntile/functions.py [hunk 15]
```diff
@@ -1467,56 +1467,61 @@ def total_sum_accum_async(
...
+                                                        dst, ignore_index)
     elif type(dst) is core_tensor.Tensor_bf16:
-        core_tensor.subtract_indexed_outputs_async_bf16(val, class_labels, dst)
+        core_tensor.subtract_indexed_outputs_async_bf16(val, class_labels,
```
Comment (Muxas) [inline wrappers/python/nntile/functions.py:1523]: single style to provide args

File: wrappers/python/nntile/layer/__init__.py [hunk 16]
```diff
@@ -32,4 +33,5 @@
 __all__ = ('Act', 'Add', 'AddSlice', 'Attention', 'AttentionSingleHead',
         'BaseLayer', 'BatchNorm2d', 'Conv2d', 'Embedding', 'FlashAttention',
         'GAP', 'LayerNorm', 'Linear', 'LlamaAttention', 'GPT2Attention',
+        'BertSelfAttention',
```
Comment (Muxas) [inline wrappers/python/nntile/layer/__init__.py:36]: no `BertAttention`?

File: wrappers/python/tests/loss/test_xentropy.py [hunk 17]
```diff
@@ -13,85 +13,117 @@
...
 import pytest
-import scipy.special as spsp
+import torch
 
 import nntile
 from nntile.loss import CrossEntropy as cross_entropy
 from nntile.tensor import TensorMoments
 
-config = nntile.starpu.Config(1, 0, 0)
+config = nntile.starpu.Config(1, 1, 0)
```
Comment (Muxas) [inline wrappers/python/tests/loss/test_xentropy.py:22]: if you ask for 1 CUDA device, then enable cublas by switching the last argument to `1`

File: wrappers/python/nntile/layer/__init__.py (hunk 16 above)
Comment (amkatrutsa) [reply wrappers/python/nntile/layer/__init__.py:36]: BertAttention is in the module subpackage since it consists of more than one layer inside

File: wrappers/python/examples/masked_lm_data_preparation.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
+import datasets
```
Comment (Muxas) [inline wrappers/python/examples/masked_lm_data_preparation.py:1]: Add license header. Even though this script is not used.

File: wrappers/python/examples/mlm_data_preparation.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+#                              (Skoltech), Russia. All rights reserved.
+#                2023-present Artificial Intelligence Research Institute
+#                              (AIRI), Russia. All rights reserved.
+#
+# NNTile is software framework for fast training of big neural networks on
+# distributed-memory heterogeneous systems based on StarPU runtime system.
+#
+# @file wrappers/python/examples/mlm_data_preparation.py
+# Data preparation for mask language modeling
+# (explicit masking appers in the training script)
```
Comment (Muxas) [inline wrappers/python/examples/mlm_data_preparation.py:11]: appeArs

File: wrappers/python/examples/roberta_training.py [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+parser = argparse.ArgumentParser(prog="Bert neural networks",
+        description="This example presents an NNTile implementation of a "
+        "Bert/RoBERTa models and compare them against the Huggingface. "
+        "It checks relative accuracy of a forward pass (values of "
+        "activations) and backward pass (gradients of parameters and "
+        "activations) and a throughput of inference and training. It can "
+        "also fine-tune a pretrained NNTile model on a chosen dataset.")
+
+parser.add_argument("--remote_model_name", type=str,
+                    default="bert-base-uncased")
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:38]: default to roberta model maybe?

File: wrappers/python/examples/roberta_training.py [hunk 21]
```diff
@@ -0,0 +1,313 @@
+# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+#                              (Skoltech), Russia. All rights reserved.
+#                2023-present Artificial Intelligence Research Institute
+#                              (AIRI), Russia. All rights reserved.
+#
+# NNTile is software framework for fast training of big neural networks on
+# distributed-memory heterogeneous systems based on StarPU runtime system.
+#
+# @file wrappers/python/examples/roberta_training.py
+# Roberta training example
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:10]: Explicitly tell a user: "RoBERTa models are based on BERT architecture, but differ by training procedure"

File: wrappers/python/examples/roberta_training.py [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+    for i in range(num_train_batches):
+        minibatch_masked_data = []
+        minibatch_labels = []
+        for j in range(num_minibatch):
+            current_mask = np.zeros((args.minibatch_size,
+                                     args.seq_len), dtype=bool)
+            idx_masked_tokens = rng.choice(args.seq_len,
+                                           size=(args.minibatch_size,
+                                                 args.n_masked_tokens_per_seq))
+            current_mask[:, idx_masked_tokens] = 1
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:232]: mask shall be different for all sequences in a batch, i think. Not a single mask for entire batch.

File: wrappers/python/examples/roberta_training.py [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+        bert_nntile, optimizer, loss, 1)
+
+# Print pipeline memory info
+pipeline.print_meminfo()
+# Warmup training
+# nntile.starpu.pause()
+nntile.starpu.profiling_enable()
+# Call separate pipeline for the single epoch for every masked data
+for i in range(args.nepochs):
+    pipeline = nntile.pipeline.Pipeline(batch_masked_data[i], batch_labels[i],
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:280]: you create a new pipeline every epoch. Old pipeline might be deleted even before execution finishes. It may lead to very bad problems. We discussed making `batch_masked_data` contain inputs for all epochs without adding a list outside. It shall be [data1_epoch1, data2_epoch1, data1_epoch2, data2_epoch2]. What is done here is the following: [[data1_epoch1, data2_epoch1], [data1_epoch2, data2_epoch2]].

**Review by Muxas — CHANGES_REQUESTED** [review]
roberta.ipynb trains bert model, not roberta

File: wrappers/python/examples/roberta_training.py [hunk 24]
```diff
@@ -244,10 +243,8 @@
             current_label[inverse_current_mask] = -100
             y.from_array(np.asfortranarray(current_label.T))
             minibatch_labels.append(y)
-        masked_data_epoch.append(minibatch_masked_data)
-        labels_epoch.append(minibatch_labels)
-    batch_masked_data.append(masked_data_epoch)
-    batch_labels.append(labels_epoch)
+    batch_masked_data.append(minibatch_masked_data)
```
Comment (Muxas) [inline wrappers/python/examples/roberta_training.py:246]: `minibatch_masked_data` is emptied on every train batch. Therefore, only the last one appears in `batch_masked_data`. Move lines 223 and 224 out of iteration over number of train batches.

Other possible solution was to switch `append` function with `extend` function in the previous version of the code:
```py
batch_masked_data.append(masked_data_epoch)
batch_labels.append(labels_epoch)
```
into
```py
batch_masked_data.extend(masked_data_epoch)
batch_labels.extend(labels_epoch)
```

File: wrappers/python/nntile/model/roberta.py [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+#                              (AIRI), Russia. All rights reserved.
+#
+# NNTile is software framework for fast training of big neural networks on
+# distributed-memory heterogeneous systems based on StarPU runtime system.
+#
+# @file wrappers/python/nntile/model/roberta.py
+# Roberta related models as a part of NNTile Python package
+#
+# @version 1.1.0
+
```
Comment (Muxas) [inline wrappers/python/nntile/model/roberta.py:13]: why not simply make RobertModel a synonim of Bert model? Here is an example:
```py
from .bert_config import BertConfigNNTile as RoBERTaConfigNNTile
...
from .bert_modules import BertModel as RoBERTaModel
```

File: wrappers/python/nntile/model/roberta_modules.py [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+from nntile.tensor import to_numpy
+
+from ..layer.act import Act
+from ..layer.layer_norm import LayerNorm
+from ..layer.linear import Linear
+from .base_model import BaseModel
+from .bert_config import BertConfigNNTile
+
+
+class RobertaLMHead(BaseModel):
```
Comment (Muxas) [inline wrappers/python/nntile/model/roberta_modules.py:29]: Is it different from BertLMHead?

**Review by Muxas — APPROVED** [review]

**Review by Muxas — APPROVED** [review]
